### PR TITLE
rpi-kernel: split into rpi, rpi2, rpi3, add rpi4

### PR DESCRIPTION
**Pull request by Piraty**
This is my second approach to the rpi4 problem (first try: https://github.com/void-linux/void-packages/pull/26000)

The problem is: how to not break existing systems while having a consistent naming scheme for rpi-kernel / rpi-base.

As ~demanded~ suggested by @the-maldridge , `rpi-kernel` continues to provide the kernel for rpi0/rpi1 and becomes a meta package that pulls `rpi{2,3}-kernel` (based on target arch) so legacy pull the new update systems.
This may be optional though and we could require users to manually install the new kernel package, possibly posting a new post on the website for that.

To support legacy systems, rpi-base is split into rpiN-base which all ship the same udev rule and depend on the respective kernel. This (still) allows to have a single entry-point to raspberry related packages.

I performed (offline) testing of the package transition process, seems fine. Script [here](https://gist.github.com/Piraty/2b0bd709c58ca2e8dd2240cc5b1141c9)

### How to
1. checkout this branch
2. build one of `rpi4-base` (aarch64*) or `rpi-base` (`armv6*`, `armv7l*` or `aarch64*`)
3. update your pi with the new packages or build fresh images: see https://github.com/void-linux/void-mklive/pull/153

---

@pbui @Duncaen @ahesford

### CODE REVIEW

**Comment by the-maldridge** [comment]
@Piraty nothing was demanded of you.  You were told the same thing as every other maintainer that if you want to make potentially hazardous changes down at the bottom layers of the stack, you are responsible for fixing breakage caused by your change.

**Comment by ahesford** [comment]
I think #26000 is a much cleaner approach. It isn't clear to me how moving `rpi-base` and `rpi-kernel` to metas that pull in their respective `rpi{1,2,3}-{base,kernel}` packages breaks image generation, but I haven't looked at it too closely. If there is a problem that I'm overlooking and this approach solves it, so be it.

If this approach is necessary, can a lot of the duplication in the `rpi*-kernel` packages be offloaded into a build style?

**Comment by Piraty** [comment]
> I think #26000 is a much cleaner approach. It isn't clear to me how moving `rpi-base` and `rpi-kernel` to metas that pull in their respective `rpi{1,2,3}-{base,kernel}` packages breaks image generation, but I haven't looked at it too closely. If there is a problem that I'm overlooking and this approach solves it, so be it.

I share your view. I'm not aware of the actual problem @the-maldridge [anticipated](https://github.com/void-linux/void-mklive/pull/153#discussion_r525026848) 
 
> If this approach is necessary, can a lot of the duplication in the `rpi*-kernel` packages be offloaded into a build style?

that's for `build_style=linux` which i worked on already (but finishing it will take its time), though this will hardly be useful for #26000 (builds all kernel variants in a big loop).
In advance of a (future) kernel build style, splitting all kernel off to their own template appears more sane (I kepts many unnecesary things in the templates on purpose so `diff srcpkgs/rpi{2,3}-kernel/template` isn't clobbered too much)

**Comment by Piraty** [comment]
This is ready for review.
* rpi4-base/rpi4-kernel are trivial and work as expected
* rpi{,2,3}-kernel / and -base need testing regarding package transition and successful boot

I did a successful transition of my rpi3 ([log here](https://gist.github.com/Piraty/bae850ed52d19e3702d2c6ac22e20196)) and would like more testers to check this.

How to:
1. build the packages: `rpi3-kernel rpi3-base rpi-kernel rpi-base`
2. copy to rpi: `rsync ... hostdir/binpkgs/<branch>/* rpi:/tmp`
3. install: `xbps-install --repository /tmp -Su` 

ping the usual suspects: @pbui @ahesford @ericonr @Duncaen @the-maldridge 
as well as interested users as of @26000:  @jsav0, @CMB , @manneorama , @LeamHall , @Anachron , @agausmann , @CameronNemo , @liketechnik

File: srcpkgs/rpi2-kernel/template [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+	# Remove unneeded architectures
+	# (save the correct one + Kconfig and delete all others)
+	mkdir -p arch-backup
+	cp -r ${hdrdest}/arch/${_arch} ${hdrdest}/arch/Kconfig arch-backup/
+	rm -rf ${hdrdest}/arch
+	mv arch-backup ${hdrdest}/arch
+	# Keep arch/x86/ras/Kconfig as it is needed by drivers/ras/Kconfig
+	mkdir -p ${hdrdest}/arch/x86/ras
+	cp -a arch/x86/ras/Kconfig ${hdrdest}/arch/x86/ras/Kconfig
```
Comment (ericonr) [inline srcpkgs/rpi2-kernel/template:231]: Shouldn't be necessary.

**Review by ericonr — COMMENTED** [review]
IMO `do_build` and `do_install` should be in a `rpi-kernel` build style, and maybe even `do_configure`.

**Comment by ericonr** [comment]
Also the kernel should be `python_version=3` by now, I think?

File: srcpkgs/rpi2-kernel/template (hunk 1 above)
Comment (Piraty) [reply srcpkgs/rpi2-kernel/template:231]: let's fix this in another patch maybe?

**Review by Duncaen — APPROVED** [review]

**Review by ahesford — APPROVED** [review]
I tested the `rpi3` and `rpi4` packages as currently split, and everything works as expected.

**Comment by Duncaen** [comment]
rpi2 upgraded fine too: https://gist.github.com/Duncaen/e6ee8a5b4cf3afdd7be14682f01dfe5e

**Comment by jsumners** [comment]
Well done @Piraty

**Pull request closed without merging**